sdk/typescript/runtime/main.go

```diff
@@ -339,18 +363,211 @@ func (t *TypescriptSdk) detectRuntime(ctx context.Context, modSource *ModuleSour
-	if err == nil {
-		if slices.Contains(entries, "package-lock.json") ||
-			slices.Contains(entries, "yarn.lock") ||
-			slices.Contains(entries, "pnpm-lock.yaml") {
-			return Node, "", nil
-		}
-		if slices.Contains(entries, "bun.lockb") {
-			return Bun, "", nil
-		}
+	if t.moduleConfig.hasFile("bun.lockb") {
+		return Bun, "", nil
+	}
+
+	if t.moduleConfig.hasFile("package-lock.json") ||
+		t.moduleConfig.hasFile("yarn.lock") ||
+		t.moduleConfig.hasFile("pnpm-lock.yaml") {
+		return Node, "", nil
 	}
 
 	// Default to node
 	return Node, "", nil
 }
+
+// detectPackageManager detects the package manager from the user's module.
+// If the package.json file has a field "packageManager", it will use that to
+// determine the package manager to use. Otherwise, it will use the default
+// package manager based on the lock files present in the module.
+//
+// If none of the above works, it will use yarn.
+//
+// Except if the package.json has an invalid value in field "packageManager", this
+// function should never return an error.
+func (t *TypescriptSdk) detectPackageManager(ctx context.Context) (SupportedPackageManager, string, error) {
+	// If the runtime is Bun, we should use BunManager
+	if t.moduleConfig.runtime == Bun {
+		return BunManager, "", nil
+	}
+
+	// If we find a package.json, we check if the packageManager is specified in `packageManager` field.
+	if t.moduleConfig.hasFile("package.json") {
+		json, err := t.moduleConfig.source.File("package.json").Contents(ctx)
+		if err != nil {
+			return "", "", fmt.Errorf("failed to read package.json: %w", err)
+		}
+
+		value := gjson.Get(json, "packageManager").String()
+		if value != "" {
+			// Retrieve the package manager and version from the value (e.g., yarn@4.2.0, pnpm@8.5.1)
+			packageManager, version, _ := strings.Cut(value, "@")
+
+			switch SupportedPackageManager(packageManager) {
+			case Yarn, Pnpm, Npm:
+				return SupportedPackageManager(packageManager), version, nil
+			default:
+				return "", "", fmt.Errorf("detected unknown package manager: %s", packageManager)
+			}
+		}
+	}
+
+	if t.moduleConfig.hasFile("bun.lockb") {
+		return BunManager, "", nil
+	}
+
+	if t.moduleConfig.hasFile("package-lock.json") {
+		return Npm, NpmDefaultVersion, nil
+	}
+
+	if t.moduleConfig.hasFile("yarn.lock") {
+		return Yarn, YarnDefaultVersion, nil
+	}
+
+	if t.moduleConfig.hasFile("pnpm-lock.yaml") {
+		return Pnpm, PnpmDefaultVersion, nil
+	}
+
+	// Default to yarn
+	return Yarn, YarnDefaultVersion, nil
+}
+
+// generateLockFile generate a lock file for the matching package manager.
+func (t *TypescriptSdk) generateLockFile(ctr *dagger.Container) (*dagger.Container, error) {
+	packageManager := t.moduleConfig.packageManager
+	version := t.moduleConfig.packageManagerVersion
+
+	switch packageManager {
+	case Yarn:
+		// Enable corepack
+		ctr = ctr.
+			WithExec([]string{"corepack", "enable"}).
+			WithExec([]string{"corepack", "use", fmt.Sprintf("yarn@%s", version)})
```

Comment:
The [standard](https://nodejs.org/api/packages.html#packagemanager) said that `packageManager` should have `<packageManager>@<version>`, so version isn't supposed to be empty.

In fact, I should catch that and throw an error in that case, because if you say `packageManager: yarn`, depending on the version you can have so many different result (between v1, 2, 3 and 4)